kfp/kfp_support_lib/src/kfp_support/workflow_support/utils/remote_jobs_utils.py

```diff
@@ -652,101 +407,132 @@ def follow_execution(
-    ) -> str:
-        """
-        This is the most simplistic transform execution parameters computation
-        :param worker_options: configuration of ray workers
-        :param actor_options: actor request requirements
-        :return: number of actors
-        """
-        import sys
-
-        from data_processing.utils import get_logger, GB
-        from kfp_support.workflow_support.utils import KFPUtils
-
-        logger = get_logger(__name__)
-
-        # convert input
-        w_options = KFPUtils.load_from_json(worker_options.replace("'", '"'))
-        a_options = KFPUtils.load_from_json(actor_options.replace("'", '"'))
-        # Compute available cluster resources
-        cluster_cpu = w_options["replicas"] * w_options["cpu"]
-        cluster_mem = w_options["replicas"] * w_options["memory"]
-        cluster_gpu = w_options["replicas"] * w_options.get("gpu", 0.0)
-        logger.info(f"Cluster available CPUs {cluster_cpu}, Memory {cluster_mem}, GPUs {cluster_gpu}")
-        # compute number of actors
-        n_actors_cpu = int(cluster_cpu * 0.85 / a_options.get("num_cpus", 0.5))
-        n_actors_memory = int(cluster_mem * 0.85 / (a_options.get("memory", GB) / GB))
-        n_actors = min(n_actors_cpu, n_actors_memory)
-        # Check if we need gpu calculations as well
-        actor_gpu = a_options.get("num_gpus", 0)
-        if actor_gpu > 0:
-            n_actors_gpu = int(cluster_gpu / actor_gpu)
-            n_actors = min(n_actors, n_actors_gpu)
-        logger.info(f"Number of actors - {n_actors}")
-        if n_actors < 1:
-            logger.warning(
-                f"Not enough cpu/gpu/memory to run transform, "
-                f"required cpu {a_options.get('num_cpus', .5)}, available {cluster_cpu}, "
-                f"required memory {a_options.get('memory', 1)}, available {cluster_mem}, "
-                f"required cpu {actor_gpu}, available {cluster_gpu}"
+    # get current namespace
+    ns = KFPUtils.get_namespace()
+    if ns == "":
+        print(f"Failed to get namespace")
+        sys.exit(1)
+    # submit job
+    remote_jobs = RayRemoteJobs(
+        server_url=server_url,
+        http_retries=additional_params.get("http_retries", 5),
+        wait_interval=additional_params.get("wait_interval", 2),
+    )
+    status, error, submission = remote_jobs.submit_job(
+        name=name, namespace=ns, request=e_params, executor=exec_script_name
+    )
+    if status != 200:
+        print(f"Failed to submit job - status: {status}, error: {error}")
+        exit(1)
+
+    print(f"submitted job successfully, submission id {submission}")
+    # print execution log
+    remote_jobs.follow_execution(
+        name=name,
+        namespace=ns,
+        submission_id=submission,
+        data_access=d_access,
+        print_timeout=additional_params.get("wait_print_tmout", 120),
+        job_ready_timeout=additional_params.get("wait_job_ready_tmout", 600),
+    )
+
+
+def execute_ray_jobs(
```

Comment:
Does it make sense to use only execute_ray_jobs function instead of having also execute_ray_job...

Reply:
Can you propose a solution? I was thinking about it, but do not quite see how to do this. Any suggestions are welcome

Reply:
the execute_ray_jobs is very similar to `execut_ray_job` the can be combined, see below:

Reply:
where is below?

Reply:
something like:
```python
def execute_ray_jobs(
    name: str,  # name of Ray cluster
    d_access: DataAccess,  # data access
    additional_params: dict[str, Any],
    e_params: dict[str, Any],
    exec_script_name: str,
    server_url: str,
) -> None:
    """
    Execute Ray jobs on a cluster periodically printing execution log. Completes when all Ray job complete.
    All of the jobs will be executed, although some of the jobs may fail.
    :param name: cluster name
    :param d_access: data access class
    :param additional_params: additional parameters for the job
    :param e_params: job execution parameters (specific for a specific transform,
                        generated by the transform workflow)
    :param exec_script_name: script to run (has to be present in the image)
    :param server_url: API server url
    :return: None
    """
    # get current namespace
    ns = KFPUtils.get_namespace()
    if ns == "":
        print(f"Failed to get namespace")
        sys.exit(1)
    # Create Ray job submitter
    remote_jobs = RayRemoteJobs(
        server_url=server_url,
        http_retries=additional_params.get("http_retries", 5),
        wait_interval=additional_params.get("wait_interval", 2),
    )
    # find config parameter
    config = None
    for key in e_params.keys():
        if key.startswith("data") and key.endswith("config"):
            config = key
            break
    single_run = False
    if config is None:
        print("Could no find config parameter")
        config_value = []
        single_run = True
    else:
        config_value = e_params[config]
        if type(config_value) is not list:
            print("config value is not a list")
            config_value = [config_value]
            single_run = True
    # remove config key from the dictionary
    launch_params = dict(e_params)
    del launch_params[config]
    # Loop through all configuration
    n_launches = 0
    for conf in config_value:
        # populate individual config and launch
        if conf is not None:
            launch_params[config] = conf
        status, error, submission = remote_jobs.submit_job(
            name=name, namespace=ns, request=launch_params, executor=exec_script_name
        )
        if status != 200:
            print(f"Failed to submit job - status: {status}, error: {error}, for configuration {conf}")
            if not single_run:
                continue
        print(f"submitted job successfully for configuration {conf}, submission id {submission}")
        # print execution log
        try:
            remote_jobs.follow_execution(
                name=name,
                namespace=ns,
                submission_id=submission,
                data_access=d_access,
                print_timeout=additional_params.get("wait_print_tmout", 120),
                job_ready_timeout=additional_params.get("wait_job_ready_tmout", 600),
            )
            n_launches += 1
        except SystemExit:
            print(f"Failed to execute job for configuration {conf}")
            continue

    if n_launches == 0:
        print("All executions failed")
        sys.exit(1)
    else:
        print(f"{n_launches} ot of {len(config_value)} succeeded")
```

Reply:
Thanks. Did it much simpler